workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourManager.cs

```diff
@@ -1,60 +1,238 @@
 using System.Collections.Generic;
+using Improbable.Gdk.Core.MonoBehaviours;
 using Improbable.Worker;
 using UnityEngine;
+using Entity = Unity.Entities.Entity;
 
 namespace Improbable.Gdk.Core
 {
-    public class ReaderWriterImpl
-    {
-        // This is placeholder code.
-    }
-
     /// <summary>
-    ///     Invokes dispatcher callbacks on GameObjects that represent entities.
+    ///     Keeps track of Reader/Writer availability for SpatialOSBehaviours on a particular GameObject and decides when
+    ///     a SpatialOSBehaviour should be enabled, calling into the SpatialOSBehaviourLibrary for injection.
     /// </summary>
-    public class SpatialOSBehaviourManager
+    internal class SpatialOSBehaviourManager
     {
-        public GameObject GameObject;
-        private Dictionary<uint, List<MonoBehaviour>> componentIdToReaderSpatialOSBehaviours;
-        private Dictionary<uint, List<MonoBehaviour>> componentIdToWriterSpatialOSBehaviours;
-        private Dictionary<MonoBehaviour, int> numUnsatisfiedComponents;
-        private List<MonoBehaviour> spatialOSBehavioursToEnable;
-        private List<MonoBehaviour> spatialOSBehavioursToDisable;
-        private Dictionary<uint, Dictionary<MonoBehaviour, ReaderWriterImpl>> readersWriters;
+        private readonly Entity entity;
+        private readonly long spatialId;
+
+        private readonly Dictionary<uint, HashSet<MonoBehaviour>> componentIdToReaderOrWriterSpatialOSBehaviours
+            = new Dictionary<uint, HashSet<MonoBehaviour>>();
+
+        private readonly Dictionary<uint, HashSet<MonoBehaviour>> componentIdToWriterSpatialOSBehaviours
+            = new Dictionary<uint, HashSet<MonoBehaviour>>();
+
+        private readonly Dictionary<MonoBehaviour, int> numUnSatisfiedReadersWriters
```

Comment:
Nit: Should be `numUnsatisifedReaderWriters`

Reply:
It has a slightly different meaning to me, though neither is really incorrect. I read "ReadersWriters" as "readers or writers", referring to the things the user is actually asking for, while a "ReaderWriters" would be talking about the implementing "ReaderWriter" objects that implement the reader and writer interfaces. Since the class is about handling what the user is asking for, I'd keep it the way it is. Or should I explicitly make it ReadersOrWriters?

Reply:
Actually I have ReaderOrWriter in another name anyway so I'll rename it to numUnsatisfiedReadersOrWriters. (Also just noticed it's capitalized as UnSatisfied atm, I might as well change that too since it's one word...)

Reply:
Yeah I was referring to the `Unsatisfied` bit - didn't notice Reader vs Readers